logic/transaction_pool.js

```diff
@@ -992,29 +993,46 @@ __private.transactionTimeOut = function(transaction) {
  * @returns {SetImmediate} error, ids[]
  */
 __private.expireTransactions = function(transactions, cb) {
-	async.eachSeries(
-		transactions,
-		(transaction, eachSeriesCb) => {
-			if (!transaction) {
-				return setImmediate(eachSeriesCb);
-			}
+	library.balancesSequence.add(
+		balancesSequenceCb => {
+			async.eachSeries(
+				transactions,
+				(transaction, eachSeriesCb) => {
+					if (!transaction) {
+						return setImmediate(eachSeriesCb);
+					}
 
-			const timeNow = Math.floor(Date.now() / 1000);
-			const timeOut = __private.transactionTimeOut(transaction);
-			// transaction.receivedAt is instance of Date
-			const seconds =
-				timeNow - Math.floor(transaction.receivedAt.getTime() / 1000);
-
-			if (seconds > timeOut) {
-				self.removeUnconfirmedTransaction(transaction.id);
-				library.logger.info(
-					`Expired transaction: ${
-						transaction.id
-					} received at: ${transaction.receivedAt.toUTCString()}`
-				);
-				return setImmediate(eachSeriesCb);
-			}
-			return setImmediate(eachSeriesCb);
+					const timeNow = Math.floor(Date.now() / 1000);
+					const timeOut = __private.transactionTimeOut(transaction);
+					// transaction.receivedAt is instance of Date
+					const seconds =
+						timeNow - Math.floor(transaction.receivedAt.getTime() / 1000);
+
+					if (seconds > timeOut) {
+						modules.transactions.undoUnconfirmed(
+							transaction,
+							undoUnconfirmErr => {
+								if (undoUnconfirmErr) {
+									library.logger.error(
+										`Failed to undo unconfirmed transaction: ${transaction.id}`,
+										undoUnconfirmErr
+									);
+									return setImmediate(cb);
```

Comment:
In case of an error on undoUnconfirm. We should throw an exception which restarts the application. Restarting the application allows us to revert changes applied to the unconfirmed state.